src/launchpad/utils/android/bundletool.py

```diff
@@ -0,0 +1,183 @@
+
+        self.bundletool_path = Path(bundletool_path)
+        if not self.bundletool_path.exists():
+            raise FileNotFoundError(f"bundletool not found at {bundletool_path}")
+
+    def _run_command(self, command: list[str], **kwargs: Any) -> tuple[int, str, str]:
+        """Run a bundletool command.
+
+        Args:
+            command: List of command arguments to pass to bundletool
+            **kwargs: Additional arguments to pass to subprocess.run
+
+        Returns:
+            Tuple of (returncode, stdout, stderr)
+
+        Raises:
+            BundletoolError: If command fails
+        """
+        cmd = [str(self.bundletool_path)] + command
+        logger.debug("Running bundletool command: %s", " ".join(cmd))
+
+        try:
+            result = subprocess.run(
+                cmd,
+                capture_output=True,
+                text=True,
+                check=False,
+                **kwargs,
+            )
+        except subprocess.SubprocessError as e:
+            raise BundletoolError(
+                f"Failed to run bundletool command: {e}",
+                -1,
+                "",
+                str(e),
+            ) from e
+
+        if result.returncode != 0:
+            raise BundletoolError(
+                f"bundletool command failed with return code {result.returncode}",
+                result.returncode,
+                result.stdout,
+                result.stderr,
+            )
+
+        return result.returncode, result.stdout, result.stderr
+
+    def build_apks(
+        self,
+        bundle_path: str | Path,
+        output_dir: str | Path,
+        device_spec: DeviceSpec,
+    ) -> None:
+        """Build APKs from an Android App Bundle.
+
+        Args:
+            bundle_path: Path to input AAB file
+            output_dir: Directory to output APKS files
+            device_spec: Device specification for APK splitting configuration.
+
+        Raises:
+            BundletoolError: If build command fails
+        """
+        temp_apks_path = create_temp_directory("apks-") / "apks.apks"
+        build_apks_command = ["build-apks", f"--bundle={bundle_path}", f"--output={temp_apks_path}"]
+
+        # Create a temporary file for the device spec JSON
+        with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False) as temp_file:
+            json.dump(device_spec.model_dump(by_alias=True), temp_file)
+            temp_device_spec_path = temp_file.name
+
+        try:
+            self._run_command(build_apks_command)
+
+            # Extract APKs for the specified device
+            extract_command = [
+                "extract-apks",
+                f"--apks={temp_apks_path}",
+                f"--output-dir={output_dir}",
+                f"--device-spec={temp_device_spec_path}",
+            ]
+            self._run_command(extract_command)
+        finally:
```

Comment:
Is this `finally` needed if everything is nested under the original `with` and `delete=True`? Not totally sure how python handles these blocks with each other.

Reply:
I think you're correct here, fixed